torch/distributed/_spmd/graph_optimization.py

```diff
@@ -0,0 +1,408 @@
+                func.__name__ not in _optimized_func
+            ), f"Cannot apply {func.__name__} twice."
+            invalid_passes = _run_before_sets[func.__name__].intersection(
+                _optimized_func
+            )
+            assert (
+                not invalid_passes
+            ), f"{invalid_passes} must run after {func.__name__}"
+
+            func(gm, *args, **kwargs)
+            gm.graph.lint()
+            gm.graph.eliminate_dead_code()
+            gm.recompile()
+            _optimized_func.add(func.__name__)
+
+            prefix = f"after_{func.__name__}"
+            if _dump_graph_folder:
+                if isinstance(gm, IterGraphModule):
+                    dump_graphs_to_files(
+                        {
+                            f"{prefix}_setup_gm": gm.setup_gm,
+                            f"{prefix}_main_gm": gm.main_gm,
+                            f"{prefix}_cleanup_gm": gm.cleanup_gm,
+                        },
+                        _dump_graph_folder,
+                    )
+                else:
+                    dump_graphs_to_files({prefix: gm}, _dump_graph_folder)
+
+        return pass_wrapper
+
+    return inner
+
+
+@dataclass(unsafe_hash=True)
+class CommBlock:
+    shape: Optional[torch.Size]
+    node_list: List[fx.Node]
+    inputs: List[fx.Node]
+    wait_nodes: List[fx.Node]
+    comm_node: fx.Node
+    outputs: Set[fx.Node]
+
+
+def get_comm_block_nodes(comm_node: fx.Node) -> CommBlock:
+    """
+    Given a wait_comm node, find out all the nodes belong to this communcation.
+
+    Args:
+        comm_node(fx.Node): The target communication/collective node.
+    Returns:
+        The CommBlock that encapsulates the related nodes (e.g., wait_node) of
+        the given comm_node.
+    """
+    node_list = []
+    wait_nodes = []
+    inputs, _ = tree_flatten((comm_node.args, comm_node.kwargs))
+    distance = 0
+    wait_prefixes = ("wait_comm", "wait_tensor")
+    non_end_users_nodes = ("split", "reshape", "getitem", "detach", "alias")
+
+    nodes = collections.deque([comm_node, None])
+    while nodes and distance < 5:
+        node = nodes.popleft()
+        if node is None:
+            distance += 1
+            if nodes:
+                nodes.append(None)
+            continue
+        node_list.append(node)
+        if node.name.startswith(wait_prefixes):
+            wait_nodes.append(node)
+        else:
+            for child in node.users:
+                if isinstance(child, fx.Node):
+                    nodes.append(child)
+
+    if not wait_nodes:
+        raise RuntimeError(
+            "The wait nodes are too far away from the comm node {comm_node}."
+        )
+
+    # Identify all the outputs
+    outputs: Dict[fx.Node, Any] = {}
+    nodes = collections.deque(wait_nodes)
+    while nodes:
+        node = nodes.popleft()
+        for user in node.users:
+            if isinstance(user, fx.Node) and user.name.startswith(non_end_users_nodes):
```

Comment:
curious, when can a user of fx.Node be non-fx.Node?

Reply:
Well, we added `__hold__` for the users of some nodes to avoid DCE in the old tracing scheme. This can probably  be removed as we no longer need to do so.

Reply:
oh I see!